### DISCUSSION TITLE
Incomplete "autocomplete" support in PyCharm

### DISCUSSION THREAD

**Discussion by klar-C** [sqlalchemy/sqlalchemy#10224] [Usage Questions]
Hi,

I'm trying to get autocomplete to work in Pycharm. Things like the below work.
![image](https://github.com/sqlalchemy/sqlalchemy/assets/51972811/a0ddcfbb-75e6-4cca-8f46-214e9c4d6f26)

However as soon as I'm trying to construct an ORM query I'm out of luck.
![image](https://github.com/sqlalchemy/sqlalchemy/assets/51972811/380e777c-86fb-45a7-b961-0896e6c9a288)

It only works when I specify the selectable type. But then I can't do method chaining.
![image](https://github.com/sqlalchemy/sqlalchemy/assets/51972811/87604d8f-baf3-4c8f-90e1-ae5056b90c52)
![image](https://github.com/sqlalchemy/sqlalchemy/assets/51972811/8f5bf549-d81c-4ed7-80e6-58fecc1f7c7b)

I've been researching quite a bit and have not yet figured out whether this is just not possible / an issue with Pycharm / me doing something wrong. I also tried it in Visual Studio Code and seems to be the same.

Any insights would be appreciated.

**Comment by zzzeek** [comment]
there have been lots of pycharm issues reported, however vscode should be working.

here's my vscode on `select()`:

![image](https://github.com/sqlalchemy/sqlalchemy/assets/128223/4221077b-d234-4836-a0a9-b7c7811c82cc)


and method chaining:

![image](https://github.com/sqlalchemy/sqlalchemy/assets/128223/3cd6e4e5-73c6-47f9-8777-59b6f5c19702)


so for vscode, make sure you are on the latest and that it does not have legacy SQLAlchemy stubs installed.    I run this script when I install vscode:

```bash
#!/bin/sh

# pyright:
# .npm/_npx/<number>/node_modules/pyright/dist/typeshed-fallback/stubs/SQLAlchemy/sqlalchemy

# pylance
# .vscode/extensions/ms-python.vscode-pylance-2021.12.3-pre.1/dist/bundled/stubs/sqlalchemy


find  ~/.vscode/ ~/.npm/ -type d -path "*/stubs*sqlalchemy" -print -exec rm -fr {} \;

```

I haven't checked but vscode really should not be shipping with legacy stubs anymore since I think typeshed has been updated.

but overall, make sure no legacy SQLAlchemy stub files are around and see if you can get vscode to work.

for pycharm, you'd have to go to them once you confirm vscode works fine

**Comment by klar-C** [comment]
Okay great thanks. I tried VSCode and same issue again.

But after further tries it looks like its a versioning issue. I was using 1.4.47 prior (needed it to be compatible with Airflow which still used 1.4).

I tried it in the most recent sqlalchemy version and there the autocomplete in Pycharm actually works.

Can you confirm that I should not expect autocomplete to work fully in 1.4?

(And to add to that but a different topic entirely: Is there any guide as to how to make sure that sqlalchemy code that was coded in 2.0 is going to work on servers that run 1.4?)

**Reply by zzzeek** [reply]
we have .pyi stubs for sqlalchemy 1.4 at https://pypi.org/project/sqlalchemy2-stubs/, which are not standard to any IDE, but if those are installed, you should get fairly OK results, but I haven't personally looked too closely at them, they were interim.   the typing situation in all SQLAlchemy versions prior to 2.0 is very poor.

**Reply by klar-C** [reply]
Thank you! That worked on a first try. I will do more testing and get back if there's any issues.

**Reply by klar-C** [reply]
So it does work but then it messes up the Autocomplete of ORM classes.

Whereas things like "MyClass.Columname." shows an autocomplete for "in_" without the stubs installed, that disappears when installing the 2.0 stubs.

**Reply by klar-C** [reply]
I guess to come back to my subquestion: Is it workable to run code that was written in 2.0 inside a container that has 1.4 installed? My undestanding always was that 1.4 has the features of 2.0 as well and that 2.0 basically just disables a lot of the older stuff?

**Reply by zzzeek** [reply]
2.0 includes many new ORM constructs that are not in 1.4, so in the general sense that is not true.  if the 2.0 code was written to carefully only make use of features that are also in 1.4 then it would work.

**Reply by klar-C** [reply]
Okay thank you. Appreciate the help!

**Comment by me-v2** [comment]
![image](https://github.com/sqlalchemy/sqlalchemy/assets/6077601/2c468c6e-c513-46a6-99c8-41c6a8a7cbb7), when instance orm, not Autocomplete  related field.
